### PR TITLE
Extend dynamicRef keyword

### PR DESCRIPTION
**Pull request by nezhar**
The description in https://stackoverflow.com/a/69774520/5382179 helped me better understand how the keyword should actually work and the PR extends the missing parts of the current implementation of the keyword.

There might be a better way to implement this, but it would require bigger changes in the way the scopes and the validator functions are handled.

### CODE REVIEW

File: jsonschema/tests/test_jsonschema_test_suite.py [hunk 1]
```diff
@@ -401,15 +401,7 @@ def leap_second(test):
...
-            message="dynamicRef support isn't working yet.",
-            subject="dynamicRef",
-        )(test)
-        or skip(
-            message="These tests depends on dynamicRef working.",
-            subject="defs",
-        )(test)
-        or skip(
-            message="These tests depends on dynamicRef working.",
+            message="These tests require an extension or the url resolver.",
```
Comment (Julian) [inline jsonschema/tests/test_jsonschema_test_suite.py:404]: Thanks! This looks promising. Can you elaborate on what you mean by an extension of the URL resolver? Looking at this test it doesn't look dynamicRef related (so I must have mis-labelled the message), but I'm not sure I see why this wouldn't work as-is with the location independent ref support you helped add.

Comment (nezhar) [reply jsonschema/tests/test_jsonschema_test_suite.py:404]: The `resolve_fragment` method located in `RefResolver` is resolving the anchor by taking the first subschema that matches the anchor name in the current schema. 

```python
for keyword in ["$anchor", "$dynamicAnchor"]:
    for subschema in self._finditem(document, keyword):
        if fragment == subschema[keyword]:
            return subschema
```

This work fine as long as the anchor names are not repeated in other scopes, this is why the test fails and resolves to:

```json
{
    "$id": "child2",
    "$anchor": "my_anchor",
    "type": "number"
}
```

We may need to extend `_finditem` to to not leave the scope while searching for the anchor or take all findings and evaluate if they not break out of the current scope before returning the result.

**Comment by Julian** [comment]
@nezhar can you have a look at resolving the conflicts here now that the performance issues with the implementation have been addressed?

**Comment by Julian** [comment]
Thanks @nezhar. I'm about ready to merge this (sorry for the delay), but to be extra safe I'd like to re-run the benchmarks from @Stranger6667 in #853 -- can you rerun the two they left at the bottom there just to be sure we don't regress on them by merging this?

**Comment by Stranger6667** [comment]
FWIW, some of the work done during ref resolving is only relevant for newer drafts (after Draft 7), so maybe as a follow-up from this PR it would be nice to explore adding some conditional logic to avoid unnecessary work (e.g. the `$dynamicAnchor` keyword is not a part of Draft 7, so could be skipped during resolving) for older drafts. Not sure how much benefit it would yield though

**Comment by nezhar** [comment]
@Julian also sorry for the delay :slightly_smiling_face:. I made another rebase and run the test from the linked issue. This are my results:

```
$ python --version; pip freeze | grep json; python test.py
Python 3.8.10
jsonschema==4.3.3
10.12309193611145 sec

$ python --version; pip freeze | grep json; python test.py
Python 3.8.10
jsonschema @ file:///home/hnezbeda/DEV/jsonschema
10.206285238265991 sec
````

**Comment by Julian** [comment]
No worries! Will merge I think. Thanks! If you manage to figure out the remaining 2 skipped tests in draft2020 a PR would of course would be very much appreciated. So long as they still fail, my suspicion is that the implementation is still not 100% correct (even in the parts that pass), but it's definitely an improvement. Thanks again!

**Pull request merged**